lib/src/rules/avoid_types_as_parameter_names.dart

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2018, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/analyzer.dart';
+
+const _desc = r'Avoid types as parameter names.';
+
+const _details = r'''
+
+**AVOID** using parameter names that is the same as an existing type.
+
+**BAD:**
+```
+m(f(int));
+```
+
+**GOOD:**
+```
+m(f(int v));
+```
+
+''';
+
+class AvoidTypesAsParameterNames extends LintRule {
+  AvoidTypesAsParameterNames()
+      : super(
+            name: 'avoid_types_as_parameter_names',
+            description: _desc,
+            details: _details,
+            group: Group.errors);
+
+  @override
+  AstVisitor getVisitor() => new Visitor(this);
+}
+
+class Visitor extends SimpleAstVisitor {
+  Visitor(this.rule);
+
+  final LintRule rule;
+
+  @override
+  visitFormalParameterList(FormalParameterList node) {
+    for (final parameter in node.parameters) {
+      if (parameter.element.hasImplicitType &&
+          (parameter.identifier.name.startsWith(new RegExp('[A-Z]')) ||
+              ['num', 'int', 'double', 'bool', 'dynamic']
+                  .contains(parameter.identifier.name))) {
```

Comment:
I haven't found how to check if the identifier name matches a type that is visible in scope. @bwilkerson how can I do that ?

Reply:
I think this will work: get the `LibraryElement` for the containing library and use `publicNamespace` to get a `Namespace`. You can ask the namespace to look up the name and then test to see whether the returned element is a `TypeDefiningElement`.

Reply:
Looks like `cu.element.library.publicNamespace.definedNames` contains only elements defined inside the library. For instance, `publicNamespace.get('int')` returns `null`.

Reply:
Oh, right. What you really want we don't expose as public API.

Internally I'd use `new LibraryScope(cu.element.library)` to get a name scope that includes all of the imported names. But that probably won't work in the future as we move analyzer on top of the front_end package, so I hesitate to suggest it. Maybe a comment about what we'd like to do here?

Reply:
Comment added.